components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +73,158 @@ impl MutationObserver {
+                    record.SetAttributeName(DOMString::from(&*name));
+                    record.SetAttributeNamespace(DOMString::from(&*namespace));
+                }
+            }
+            //Step 4.3-4.6- TODO currently not relevant to mutation records for attribute mutations
+            //Step 4.7
+            if pairedStrings[i] != "" {
+                record.SetoldValue(pairedStrings[i].clone());
+            }
+            i = i + 1;
+        }
+        // Step 5
+        MutationObserver::queueMutationObserverCompoundMicrotask();
+    }
+
+}
+
+impl MutationObserverMethods for MutationObserver {
+    /// https://dom.spec.whatwg.org/#dom-mutationobserver-observe
+    /// MutationObserver.observe method
+    fn Observe(&self, target: &Node, options: &MutationObserverInit) -> Fallible<()> {
+        // Step 1: If either options’ attributeOldValue or attributeFilter is present and
+        // options’ attributes is omitted, set options’ attributes to true.
+        if (options.attributeOldValue.is_some() || options.attributeFilter.is_some()) && options.attributes.is_none() {
+//            options.attributes.set(Some(true));
+        }
+        // Step2: If options’ characterDataOldValue is present and options’ characterData is omitted,
+        // set options’ characterData to true.
+        if options.characterDataOldValue.is_some() && options.characterData.is_some() {
+//            options.characterData.set(Some(true));
+        }
+        // Step3: If none of options’ childList, attributes, and characterData is true, throw a TypeError.
+        if !options.childList && !options.attributes.unwrap_or(false) && !options.characterData.unwrap_or(false) {
+            return Err(Error::Type("childList, attributes, and characterData not true".to_owned()));
+        }
+        // Step4: If options’ attributeOldValue is true and options’ attributes is false, throw a TypeError.
+        if options.attributeOldValue.unwrap_or(false) && !options.attributes.unwrap_or(false) {
+            return Err(Error::Type("attributeOldValue is true but attributes is false".to_owned()));
+        }
+        // Step5: If options’ attributeFilter is present and options’ attributes is false, throw a TypeError.
+        if options.attributeFilter.is_some() && !options.attributes.unwrap_or(false) {
+            return Err(Error::Type("attributeFilter is present but attributes is false".to_owned()));
+        }
+        // Step6: If options’ characterDataOldValue is true and options’ characterData is false, throw a TypeError.
+        if options.characterDataOldValue.unwrap_or(false) && !options.characterData.unwrap_or(false) {
+            return Err(Error::Type("characterDataOldValue is true but characterData is false".to_owned()));
+        }
+        // Step 7
+        for registered in target.registered_mutation_observers().borrow().iter() {
+//            if &*registered as *const Root<MutationObserver> == self as *const MutationObserver {
```

Comment:
This check needs to be reinstated - maybe `&**registered` will allow `as *const MutationObserver`?